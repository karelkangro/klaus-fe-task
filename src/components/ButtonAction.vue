<template>
  <button :class="['button', { 'button--square': square }]">
    <slot />
    {{ name }}
  </button>
</template>

<script>
export default {
  name: "button-action",
  props: {
    name: String,
    square: Boolean
  }
};
</script>

<style lang="scss">
.button {
  width: auto;
  height: var(--space-2);
  margin-right: var(--space-0-25);
  display: flex;
  align-items: center;
  background: white;
  color: var(gray-70);
  font-weight: 500;
  font-size: var(--text-sm);
  border: 0.0625rem solid var(--gray-30);
  border-radius: var(--border-radius-4);
  padding-right: var(--space-1);
  box-shadow: var(--shadow-xs);

  svg {
    margin-right: var(--space-0-5);
  }

  &--square {
    margin: 0;
    padding: var(--space-0-5);

    svg {
      margin: 0;
    }
  }
}
</style>
