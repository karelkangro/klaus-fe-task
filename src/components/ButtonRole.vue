<template>
  <button
    :class="[
      'button-role',
      { 'button-role--admin': role === 'ADMIN' },
      { 'button-role--agent': role === 'AGENT' },
      { 'button-role--account': role === 'ACCOUNT_MANAGER' },
      { 'button-role--external': role === 'EXTERNAL_REVIEWER' }
    ]"
  >
    {{ parsedRole }}
  </button>
</template>

<script>
import { parsedString } from "@/utils";

export default {
  name: "button-role",
  props: {
    role: String
  },
  computed: {
    parsedRole() {
      return parsedString(this.role);
    }
  }
};
</script>

<style lang="scss">
.button-role {
  border: none;
  border-radius: var(--border-radius-4);
  height: 1.5rem;
  padding-left: var(--space-0-5);
  padding-right: var(--space-0-5);
  font-size: var(--text-xs);

  &--admin {
    background: var(--purple-20);
  }
  &--agent {
    background: var(--blue-20);
  }
  &--account {
    background: var(--pink-20);
  }
  &--external {
    background: var(--orange-20);
  }
}
</style>
