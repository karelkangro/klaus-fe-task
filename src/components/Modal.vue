<template>
  <div class="modal" v-if="isOpen">
    <div class="modal__content">
      <div class="modal__body">
        <slot />
      </div>
      <div class="modal__footer">
        <button @click="closeModal">
          {{ closeButtonLabel }}
        </button>
      </div>
    </div>
  </div>
</template>
<script>
import { watch } from "vue";

export default {
  name: "modal",
  props: {
    isOpen: {
      type: Boolean,
      default: false
    },
    closeButtonLabel: {
      type: String
    }
  },
  setup(props, { emit }) {
    const closeModal = () => emit("close-modal");

    const toggleBodyOverflowHidden = () => {
      if (props.isOpen) document.body.style.overflow = "hidden";
      if (!props.isOpen) document.body.style.overflow = "";
    };

    watch(
      () => props.isOpen,
      () => toggleBodyOverflowHidden()
    );

    return {
      closeModal
    };
  }
};
</script>
<style lang="scss">
.modal {
  position: fixed;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  background: rgba(0, 0, 0, 0.6);
  z-index: 3;

  &__content {
    display: block;
    width: 35rem;
    max-height: 100%;
    padding-bottom: var(--space-2);
    position: fixed;
    left: 50%;
    top: 50%;
    transform: translate(-50%, -50%);
    background: white;
    box-shadow: 0 0 3.75rem 0.625rem rgba(0, 0, 0, 0.9);
    border-radius: var(--border-radius-8);
    border: 0.0625rem solid var(--color-gray);
    overflow-x: scroll;
    z-index: 2;
  }

  &__body {
    padding: var(--space-3);
  }

  &__footer {
    position: relative;
    width: 100%;
    margin-left: auto;
    bottom: var(--space-1);
    display: flex;

    button {
      font-size: var(--text-sm);
      font-weight: 500;
      background: var(--brand-50);
      min-width: 2rem;
      color: white;
      border: none;
      border-radius: var(--border-radius-4);
      padding: var(--space-1);
      margin-left: auto;
      margin-right: var(--space-1-5);
      z-index: 1;
    }
  }
}
</style>
