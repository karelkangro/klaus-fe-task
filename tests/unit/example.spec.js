import { shallowMount } from "@vue/test-utils";
import ButtonAction from "@/components/ButtonAction.vue";
import ButtonRole from "@/components/ButtonRole.vue";
import Modal from "@/components/Modal.vue";

describe("ButtonAction.vue", () => {
  it("renders name prop when passed", () => {
    const name = "Send";
    const wrapper = shallowMount(ButtonAction, {
      props: { name }
    });
    expect(wrapper.text()).toMatch(name);
  });

  it("renders props.msg when passed", () => {
    const wrapper = shallowMount(ButtonAction, {
      props: { square: true }
    });
    expect(wrapper.classes()).toContain("button--square");
  });
});

describe("ButtonRole.vue", () => {
  const roles = [
    {
      name: "ADMIN",
      class: "button-role--admin"
    },
    {
      name: "AGENT",
      class: "button-role--agent"
    },
    {
      name: "ACCOUNT_MANAGER",
      class: "button-role--account"
    },
    {
      name: "EXTERNAL_REVIEWER",
      class: "button-role--external"
    }
  ];

  roles.forEach(role => {
    it(`renders role prop ${role.name} when passed`, () => {
      const wrapper = shallowMount(ButtonRole, {
        props: { role: role.name }
      });
      expect(wrapper.classes()[1]).toEqual(role.class);
    });
  });
});

describe("Modal.vue", () => {
  it("renders nothing when isOpen prop is false", () => {
    const wrapper = shallowMount(Modal, {
      props: { isOpen: false }
    });
    expect(wrapper.html()).toBe("<!--v-if-->");
  });

  it("renders modal when isOpen prop is true", () => {
    const wrapper = shallowMount(Modal, {
      props: { isOpen: true }
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("renders closeButtonLabel prop when passed", () => {
    const name = "Send";
    const wrapper = shallowMount(Modal, {
      props: { closeButtonLabel: name, isOpen: true }
    });
    expect(wrapper.find("button").text()).toEqual(name);
  });

  it("emits close-modal event when close button is pressed", async () => {
    const name = "Send";
    const closeModal = jest.fn();
    const wrapper = shallowMount(Modal, {
      props: { closeButtonLabel: name, isOpen: true, closeModal }
    });
    await wrapper.find("button").trigger("click");
    expect(wrapper.emitted("close-modal")).toHaveLength(1);
  });
});
