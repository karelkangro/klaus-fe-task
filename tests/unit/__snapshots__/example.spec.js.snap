// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal.vue renders modal when isOpen prop is true 1`] = `
<div class="modal">
  <div class="modal__content">
    <div class="modal__body"></div>
    <div class="modal__footer"><button></button></div>
  </div>
</div>
`;

exports[`Modal.vue renders modal when isOpen prop is true 2`] = `
<div class="modal">
  <div class="modal__content">
    <div class="modal__body"></div>
    <div class="modal__footer"><button></button></div>
  </div>
</div>
`;
